Add ink! linting MVP stage

cmichi opened the pull request:
This is an MVP, there are a couple shortcomings in there for which I want to create follow-ups:

* As it is now we unfortunately have to require people to install `dylint` manually. We should look into making it a proper dependency instead. However, this was not straight forward.
* We should render and publish a proper website displaying info about the lint (the one which right now is just in the doc comment). So similar to https://rust-lang.github.io/rust-clippy/master/.
* We should recognize nested calls to `ink_lang::codegen::initialize_contract`. So if it's not called directly in the constructor, but instead in a sub-function.

xgreenx commented:
> We should recognize if the call to ink_lang::codegen::initialize_contract is nested within the constructor. So if it's not called directly in the constructor, but instead in a sub-function.

Maybe better to require the user to always implement `SpreadAllocate` and we will use `ink_lang::codegen::initialize_contract` in the codegen?

If the contract implements `SpreadAllocate` we can create an instance of the contract. After we can pass that instance into the constructor and the user will do all the necessary stuff.

[Example](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=52cb2e5f10e7005478cd7a8a36ebca80) of the idea in the playground. 

```rust
impl Contract {
    #[ink(constrcutor)]
    fn new(mut self/* we will require `self` here*/, total_supply: u32) -> Self {
        self.total_supply = total_supply;
        self
    }
}
```

cmichi commented:
@xgreenx Thanks for your comment! So the thing is that we don't want to change the ink! constructor function signature, we would really like to stay as close to Rust syntax as possible.

Putting the `initialize_contract` call into the codegen would result in a changed function signature though. This has implications for e.g. the off-chain testing, so there some special API be needed so that it's still possible to e.g. write Rust unit tests for the contract.

Our thinking right now is that we will likely find a better way going forward and by requiring the `initalize_contract` in the meantime those contracts would be forward compatible. If we change the function signature now though, then those contracts can't be easily used with newer versions of ink! if we change the function signature back.

xgreenx commented:
Thanks for the response. I see you are right. I forgot about off-chain unit tests that the problem still exists for them.

> Our thinking right now is that we will likely find a better way going forward and by requiring the initalize_contract in the meantime those contracts would be forward compatible. If we change the function signature now though, then those contracts can't be easily used with newer versions of ink! if we change the function signature back.

Yesterday I tried to implement your idea regarding the new structure of the contract. [Here](https://github.com/paritytech/ink/issues/1134#issuecomment-1042351934) is results. Maybe you can check it in your free time=)

sergejparity commented on .github/workflows/macos.yml:
```diff
@@ -26,7 +26,7 @@ jobs:
 
       - name: Checkout sources & submodules
-        uses: actions/checkout@master
+        uses: actions/checkout@v2.4.0
```
```suggestion
        uses: actions/checkout@v2
```
It is recommended to use this tag as it following updates to the latest minor version

sergejparity commented on .github/workflows/windows.yml:
```diff
@@ -32,27 +32,40 @@ jobs:
 
       - name: Checkout sources & submodules
-        uses: actions/checkout@master
+        uses: actions/checkout@v2.4.0
```
```suggestion
        uses: actions/checkout@v2
```
It is recommended to use this tag as it following updates to the latest minor version

sergejparity commented on .github/workflows/macos.yml:
```diff
@@ -36,20 +36,33 @@ jobs:
       - name: Install toolchain
         id: toolchain
-        uses: actions-rs/toolchain@master
+        uses: actions-rs/toolchain@v1.0.7
```
```suggestion
        uses: actions-rs/toolchain@v1
```
It is recommended to use this tag as it following updates to the latest minor version

sergejparity commented on .github/workflows/windows.yml:
```diff
@@ -32,27 +32,40 @@ jobs:
       - name: Install toolchain
         id: toolchain
-        uses: actions-rs/toolchain@master
+        uses: actions-rs/toolchain@v1.0.7
```
```suggestion
        uses: actions-rs/toolchain@v1
```
It is recommended to use this tag as it following updates to the latest minor version

ascjones commented on build.rs:
```diff
@@ -42,36 +42,147 @@ fn main() {
+        // capture the stdout to return from this function as bytes
+        .stdout(std::process::Stdio::piped())
+        .spawn()
+        .unwrap_or_else(|_| panic!("Error executing `{:?}`", cmd));
```
Return `Err` instead of panicking here?

ascjones commented on build.rs:
```diff
@@ -42,36 +42,147 @@ fn main() {
+        let zipped_lints = build_dylint_driver(manifest_dir, out_dir, dylint_driver_dst_file)
+            .map(|_| true)
+            .unwrap_or_else(|err| {
+                panic!("Failed building dylint driver: {:?}", err);
```
If we are going to exit cleanly with an exit code below, then to be consistent it should either handle results or do `eprintln!` + `exit(1)`.

Or instead we can just `panic!` everywhere which might have the same result. Either way it should be consistent.

ascjones commented on build.rs:
```diff
@@ -42,36 +42,147 @@ fn main() {
+            .unwrap_or_else(|err| {
+                panic!("Failed building dylint driver: {:?}", err);
+            });
+        std::process::exit(bool_to_exit_code(zipped_template && zipped_lints));
```
I would just replace this whole block with a call to `build_dylint_driver` and then add an alternative with `#[cfg(feature = "cargo-clippy")]` since you already have that method decorated. It will make the control flow easier to follow

ascjones commented on .gitlab-ci.yml:
```diff
@@ -81,6 +87,26 @@ clippy:
+    - cargo fmt --verbose --all -- --check
+    - cargo clippy --verbose -- -D warnings;
+
+    - cargo test --verbose --workspace --all-features
```
Is the `--workspace` necessary? Seems to run without it. Harmless anyway

athei replied:
Yes. Because we are not using a virtual manifest. All commands only run on the main package in thise case without `--workspace` or a `default-member` key in the manifest.

ascjones commented on ink_linting/ui/pass/mapping-two-constructors.rs:
```diff
@@ -0,0 +1,60 @@
+    }
+
+    impl MyContract {
+        /// The `initialize_contract` must be found, even if there is logic before it.
```
Is this a copy of the "additional logic" UI test? There is only a single constructor here where there should be two according to the test name.

ascjones commented on build.rs:
```diff
@@ -42,36 +42,151 @@ fn main() {
+}
+
+/// Returns the process exit code which corresponds to a boolean.
+fn bool_to_exit_code(val: bool) -> i32 {
```
This function only called from the one place now so could be moved inline

ascjones commented on build.rs:
```diff
@@ -42,36 +42,151 @@ fn main() {
+            eprintln!("Error: {:?}", e);
+            Err(e)
+        }
+    }
```
To avoid destructuring the result here, could add another high level method below main which does roughly:

```
fn zip_template_and_build_dylint_driver() -> Result<()> {
  zip_template()?;
  build_dylint_driver()?;
  println!("Some success message which used to be in the Ok arm");
  Ok(())
}
```
Then at the top level in `main()` you can do the `eprintln!` which will handle all errors, and only need to destructure once. The `println!`s in the `Ok` arm can just move inline to the methods.

athei commented on src/cmd/build.rs:
```diff
@@ -298,8 +295,88 @@ fn exec_cargo_for_wasm_target(
+            .bright_yellow());
+    }
+
+    if !execute_cmd(Command::new("dylint-link").arg("--version")) {
+        anyhow::bail!("dylint-link was not found!\n\
+            Make sure it is installed and the binary is in your PATH environment.\n\n\
+            You can install it by executing `cargo install dylint-link`."
+            .to_string()
+            .bright_yellow());
```
I think this is useless here? We don't need `dylint-link` when running `cargo-contract` but when building it. This should go into `./build.rs` in order to give a proper error when this isn't installed during build.

ascjones commented on ink_linting/src/mapping_initialized.rs:
```diff
@@ -0,0 +1,345 @@
+                MAPPING_INITIALIZED,
+                item.span,
+                &format!(
+                    "you have declared an `#[ink(storage)]` on `{}` without initializing it in the contract constructor.",
```
Needs to be more specific, and perhaps without naming and shaming the user. If you look at the Rust compiler errors they are normally in the passive voice https://doc.rust-lang.org/error-index.html.

Here's a few ideas (though try and come up with something more readable!)

"missing contract initialization in constructor for #[ink(storage)] struct with a Mapping field."
"Storage struct with a Mapping field must be initialized on construction"

ascjones commented on ink_linting/src/mapping_initialized.rs:
```diff
@@ -0,0 +1,345 @@
+/// call `initialize_contract` within a sub-function of the ink! constructor
+/// this is not recognized!
+///
+/// Also the function call must be explicitly called `ink_lang::utils::initialize_contract(…)`,
```
Is this true though? I modified the `mapping_one_constructor` to `use ink_lang::utils::initialize_contract;` and it still works. Also if you look at the regex it uses the non reexported name to match on

ascjones reviewed (commented):
First pass complete

cmichi commented on .gitlab-ci.yml:
```diff
@@ -81,6 +87,26 @@ clippy:
 
 #### stage:                        test (all features)
 
+test-dylint:
+  stage:                           test
+  <<:                              *docker-env
+  script:
```
Yes exactly, rustc starts complaining about `incompatible metadata version found`.

cmichi commented:
@ascjones @athei I've addressed your comments and updated the PR. The error which Alex encountered happened when no `--manifest-path` was supplied, I've fixed that as well.

athei commented on src/cmd/build.rs:
```diff
@@ -298,8 +295,108 @@ fn exec_cargo_for_wasm_target(
+    let tmp_dir_path = tmp_dir.path().as_os_str().to_string_lossy();
+    let env = vec![
+        ("DYLINT_LIBRARY_PATH", Some(tmp_dir_path.as_ref())),
+        ("DYLINT_DRIVER_PATH", Some(tmp_dir_path.as_ref())),
```
When I remove this line the driver is cached and doesn't need to be rebuild on every check. Makes sense because you are compiling it to a new temporary directory each time.

I don't know where it is put when I remove this line, though.

athei commented on src/cmd/build.rs:
```diff
@@ -298,8 +295,108 @@ fn exec_cargo_for_wasm_target(
+    crate::util::unzip(template, tmp_dir.path().to_path_buf(), None)?;
+
+    let manifest_path = crate_metadata.manifest_path.cargo_arg()?;
+    let args = vec!["ink_linting", &manifest_path];
```
Let's be explicit.

```suggestion
    let args = vec!["--lib", "ink_linting", &manifest_path];
```

athei commented on src/cmd/build.rs:
```diff
@@ -298,8 +295,108 @@ fn exec_cargo_for_wasm_target(
+        .tempdir()?;
+    log::debug!("Using temp workspace at '{}'", tmp_dir.path().display());
+
+    let template = include_bytes!(concat!(env!("OUT_DIR"), "/ink-dylint-driver.zip"));
```
Copy paste variable name?

athei replied:
I mean replacing `let template` by `let driver` or something.

ascjones commented on build.rs:
```diff
@@ -35,43 +35,160 @@ use substrate_build_script_utils::rerun_if_git_head_changed;
-        dst_file.display()
+        template_dst_file.display()
     );
+    zip_dir(&template_dir, &template_dst_file, CompressionMethod::Stored).map(|_| {
```
Not a fan of `map`ping into an io side effect, but now I am just nitpicking

ascjones reviewed (approved):
LGTM

athei commented:
I am wondering: We are writing our custom link but don't even run the stock clippy lints. Can we do that at the same time as we are running dylint so we don't have multiple passes?

xgreenx commented:
Is that possible to use `CARGO_TARGET_DIR` with `dylint` to not build it each time?
Also maybe make `dylint` configurable? For example, I don't want to build it during integration tests=)
